src/main/java/io.cdap.plugin/Flatten.java

```diff
@@ -0,0 +1,511 @@
+    collector.getOrThrowException();
+
+    List<OutputFieldInfo> inputOutputMapping = new ArrayList<>();
+    try {
+      inputOutputMapping = createOutputFieldsInfo(inputSchema, fieldsToFlatten,
+                                                  config.getLevelToLimitFlattening());
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+
+    Map<String, OutputFieldInfo> outputFieldMap = handleDuplicateFieldName(inputOutputMapping, config.prefix,
+                                                                           collector);
+    Schema outputSchema = generateOutputSchema(inputSchema, outputFieldMap.values());
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
+  }
+
+  private void checkMaxLimitForNestedLevels(Integer levelToLimitFlattening, FailureCollector collector) {
+    if (levelToLimitFlattening != null && levelToLimitFlattening > MAX_NESTED_LEVEL) {
+      collector.addFailure(String.format("Maximum level for nested is %d.", MAX_NESTED_LEVEL), null);
+      collector.getOrThrowException();
+    }
+  }
+
+  private void checkSchemaType(Schema inputSchema, List<String> fieldsToFlatten, FailureCollector collector) {
+    for (String field : fieldsToFlatten) {
+      Schema schemaField = inputSchema.getField(field).getSchema();
+      if (!isRecord(schemaField)) {
+        collector.addFailure(String.format("'%s' cannot be flattened.", field),
+                             "Only fields with schema type of `record` can be flattened.");
+      }
+    }
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+
+    failureCollector = context.getFailureCollector();
+
+    List<String> fieldsToFlatten = config.getFieldsToFlatten();
+    if (fieldsToFlatten == null || fieldsToFlatten.isEmpty()) {
+      failureCollector.addFailure("Atleast one field should be selected.", "");
+      failureCollector.getOrThrowException();
+    }
+
+    Schema inputSchema = context.getInputSchema();
+    if (inputSchema == null) {
+      return;
+    }
+
+    checkMaxLimitForNestedLevels(config.getLevelToLimitFlattening(), failureCollector);
+    List<OutputFieldInfo> inputOutputMapping = createOutputFieldsInfo(inputSchema, fieldsToFlatten,
+                                                                      config.getLevelToLimitFlattening());
+    this.inputOutputMapping = handleDuplicateFieldName(inputOutputMapping, config.getPrefix(), failureCollector);
+    this.outputSchema = generateOutputSchema(inputSchema, inputOutputMapping);
+  }
+
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    if (outputSchema == null) {
+      checkMaxLimitForNestedLevels(config.getLevelToLimitFlattening(), failureCollector);
+      List<OutputFieldInfo> inputOutputMapping = createOutputFieldsInfo(input.getSchema(), config.getFieldsToFlatten(),
```

Comment:
Is the same class instance used to transform multiple input records? If yes then it seems we are creating inputOutputMapping only for the first record.

Reply:
`OutputSchema` is created during `initialize` stage except when `inputSchema` is macro and OutputSchema should be created only once, so if `inputSchema` its macro we generate the outputSchema once and save it in so we do need to regenerate outputSchema for every record. Also the local variable `inputOutputMapping` is renamed.